android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMHardwareKeyboardInterpreter.java

```diff
@@ -114,18 +116,70 @@ public KMHardwareKeyboardInterpreter(Context context, KMManager.KeyboardType key
 
   @Override
   public boolean onKeyDown(int keyCode, KeyEvent event) {
+    // Note: Keep this table in sync with kmwosk.js osk.modifierCodes
+    final HashMap<String, Integer> modifierCodes = new HashMap<String, Integer>() {{
+      put("LCTRL", 0x0001);
+      put("RCTRL", 0x0002);
+      put("LALT", 0x0004);
+      put("RALT", 0x0008);
+      put("SHIFT", 0x0010);
+      put("CTRL", 0x0020);
+      put("ALT", 0x0040);
+      put("CAPS", 0x0100);
+      put("NO_CAPS", 0x0200);
+      put("NUM_LOCK", 0x0400);
+      put("NO_NUM_LOCK", 0x0800);
+      put("SCROLL_LOCK", 0x1000);
+      put("NO_SCROLL_LOCK", 0x2000);
+      put("VIRTUAL_KEY", 0x4000);
+    }};
+
     if (keyCode > 84 || keyCode < 0) {
       // The key is outside the range of keys we understand
       return false;
     }
 
+    boolean isChiral = false;
+    if (this.keyboardType == KMManager.KeyboardType.KEYBOARD_TYPE_INAPP) {
+      isChiral = KMManager.InAppKeyboard.getChirality();
+    } else if (this.keyboardType == KMManager.KeyboardType.KEYBOARD_TYPE_SYSTEM) {
+      isChiral = KMManager.SystemKeyboard.getChirality();
+    }
+
+    // States of modifier keys
+    // KeyEvent.getModifiers() specifically masks out lock keys (KeyEvent.META_CAPS_LOCK_ON,
+    // KeyEvent.META_SCROLL_LOCK_ON, KeyEvent.META_NUM_LOCK_ON), so get their states separately
     int androidModifiers = event.getModifiers(), keymanModifiers = 0;
-    if ((androidModifiers & KeyEvent.META_SHIFT_ON) != 0) keymanModifiers |= 0x10;
-    if ((androidModifiers & KeyEvent.META_CTRL_ON) != 0) keymanModifiers |= 0x20;
-    if ((androidModifiers & KeyEvent.META_ALT_ON) != 0) keymanModifiers |= 0x40;
+    boolean capsOn = event.isCapsLockOn();
+    boolean numOn = event.isNumLockOn();
+    boolean scrollOn = event.isScrollLockOn();
+
+    // By design, SHIFT is non-chiral
+    if ((androidModifiers & KeyEvent.META_SHIFT_ON) != 0) {
+      keymanModifiers |= modifierCodes.get("SHIFT");
+    }
+    if ((androidModifiers & KeyEvent.META_CTRL_LEFT_ON) != 0) {
+      keymanModifiers |= isChiral ? modifierCodes.get("LCTRL") : modifierCodes.get("CTRL");
+    }
+    if ((androidModifiers & KeyEvent.META_CTRL_RIGHT_ON) != 0) {
+      keymanModifiers |= isChiral ? modifierCodes.get("RCTRL") : modifierCodes.get("CTRL");
+    }
+    if ((androidModifiers & KeyEvent.META_ALT_LEFT_ON) != 0) {
+      keymanModifiers |= isChiral ? modifierCodes.get("LALT") : modifierCodes.get("ALT");
+    }
+    if ((androidModifiers & KeyEvent.META_ALT_RIGHT_ON) != 0) {
+      keymanModifiers |= isChiral ? modifierCodes.get("RALT") : modifierCodes.get("ALT");
+    }
+
+    int Lstates = 0;
+    if (isChiral) {
+      Lstates |= capsOn ? modifierCodes.get("CAPS") : modifierCodes.get("NO_CAPS");
+      Lstates |= numOn ? modifierCodes.get("NUM_LOCK") : modifierCodes.get("NO_NUM_LOCK");
+      Lstates |= scrollOn ? modifierCodes.get("SCROLL_LOCK") : modifierCodes.get("NO_SCROLL_LOCK");
```

Comment:
I don't think it's safe to predicate this test on the keyboard being chiral.  It'd be reasonable for a KMW 10.0 compatible keyboard to not use chiral information but still use Caps Lock.  Actually, a condition check shouldn't be at all necessary here.

If you're worried about Lstates causing issues with older keyboards, don't be - they're not broken by the inclusion of this information... and if they are, that's a problem to be solved within KMW, not here.